README.md
# observable-backoff [![Build Status](https://travis-ci.org/tjmehta/observable-backoff.svg?branch=master)](https://travis-ci.org/tjmehta/observable-backoff) [![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat)](http://standardjs.com/)
RxJS observable exponential backoff operator

# Installation
```bash
npm i --save observable-backoff
npm i --save rxjs # peer dependency
```

# Usage
Example: exponential backoff
```js
var Observable = require('rxjs/Observable').Observable
var Subscription = require('rxjs/Subscription').Subscription
require('observable-backoff')
var trace = function (msg) {
  console.log('trace', '-', Date.now(), '-', msg)
}

// create an observable..
var timerObservable = new Observable(function (observer) {
  trace('subscribe')
  var count = 0
  // setup timer counter
  var interval = setInterval(function () {
    observer.next(count++)
  }, 1000)
  // setup a hard-coded error after 3.5 sec..
  setTimeout(function () {
    observer.error(new Error('boom'))
  }, 3500)
  // return subscription
  return new Subscription(function () {
    trace('unsubscribe')
    clearInterval(interval)
  })
})

// make it exponentially backoff on error
var opts = {
  minTimeout: 1000,
  maxTimeout: 3000,
  factor: 3,
  retries: 2
  onError: function (err, isFinal, count) {
    // you could log every error here
  }
}
timerObservable = timerObservable.backoff(opts)

// subscribe, start timer
timerObservable(
  function (next) {
    trace('next: ' + next)
  },
  function (err) {
    trace('error: ' + err)
  }
  function () {
    trace('completed')
  }
)

/** LOG
trace - 1469418557021 - subscribe
trace - 1469418558029 - next: 0
trace - 1469418559034 - next: 1
trace - 1469418560037 - next: 2
trace - 1469418560529 - unsubscribe
// 1000ms delay before retry
trace - 1469418561538 - subscribe
trace - 1469418562544 - next: 0
trace - 1469418563550 - next: 1
trace - 1469418564554 - next: 2
trace - 1469418565043 - unsubscribe
// 3000ms delay before retry
trace - 1469418568049 - subscribe
trace - 1469418569052 - next: 0
trace - 1469418570057 - next: 1
trace - 1469418571061 - next: 2
trace - 1469418571553 - unsubscribe
trace - 1469418571554 - error: Error: boom
*/
```

# License
MIT


CHANGELOG.md
# Changelog

## 2.0.0
* Breaking: stop retrying if opts.onError returns a truthy value


test/observable-backoff.test.js
'use strict'

var noop = require('101/noop')
var Observable = require('rxjs/Observable').Observable
var Subscription = require('rxjs/Subscription').Subscription
var sinon = require('sinon')
var StaticObservable = require('static-observable')

require('../index.js')

var afterEach = global.afterEach
var beforeEach = global.beforeEach
var describe = global.describe
var it = global.it

describe('observable-backoff', function () {
  beforeEach(function () {
    this.clock = sinon.useFakeTimers()
  })
  afterEach(function () {
    this.clock.restore()
  })

  describe('error observable', function () {
    beforeEach(function () {
      var self = this
      this.err = new Error('boom')
      // observable
      this.unsubscribe = sinon.spy(noop)
      this.subscribe = sinon.spy(function (observer) {
        setTimeout(function () {
          observer.error(self.err)
        }, 0)
        return new Subscription(self.unsubscribe)
      })
      this.observable = new Observable(this.subscribe)
      // callbacks
      this.onNext = sinon.stub()
      this.onError = sinon.stub()
      this.onCompleted = sinon.stub()
    })

    it('should retry w/ exponential backoff', function () {
      var self = this
      var opts = {
        minTimeout: 1000,
        maxTimeout: 3000,
        factor: 3,
        retries: 2,
        onError: sinon.stub()
      }
      this.observable.backoff(opts).subscribe(
        this.onNext,
        this.onError,
        this.onCompleted
      )
      // observable assertions
      sinon.assert.calledOnce(this.subscribe)
      sinon.assert.notCalled(this.unsubscribe)
      this.clock.tick(1)
      sinon.assert.calledOnce(self.unsubscribe)
      sinon.assert.calledOnce(opts.onError)
      sinon.assert.calledWith(opts.onError, this.err, false, 0)
      this.clock.tick(1000)
      sinon.assert.calledTwice(self.subscribe)
      this.clock.tick(1)
      sinon.assert.calledTwice(self.unsubscribe)
      sinon.assert.calledTwice(opts.onError)
      sinon.assert.calledWith(opts.onError, this.err, false, 1)
      this.clock.tick(1000)
      sinon.assert.calledTwice(self.subscribe)
      this.clock.tick(1000)
      sinon.assert.calledTwice(self.subscribe)
      this.clock.tick(1000)
      sinon.assert.calledThrice(self.subscribe)
      this.clock.tick(1)
      sinon.assert.calledThrice(self.unsubscribe)
      sinon.assert.calledThrice(opts.onError)
      sinon.assert.calledWith(opts.onError, this.err, true, 2)
      // callback assertions
      sinon.assert.notCalled(this.onNext)
      sinon.assert.notCalled(this.onCompleted)
      sinon.assert.calledOnce(this.onError)
      sinon.assert.calledWith(this.onError, this.err)
    })
  })
})

__tests__/observable-backoff.test.js
/* eslint-env jest */
const Observable = require('rxjs/Observable').Observable
const Subscription = require('rxjs/Subscription').Subscription
const StaticObservable = require('static-observable')
const timeout = require('timeout-then')
require('observable-backoff')

describe('observable-backoff', () => {
  describe('no errors', () => {
    it('should work like a normal observable', () => {
      const observable = new StaticObservable()
      const callbacks = {
        onNext: jest.fn(),
        onError: jest.fn(),
        onComplete: jest.fn()
      }
      observable
        .backoff()
        .subscribe(
          callbacks.onNext,
          callbacks.onError,
          callbacks.onComplete
        )
      observable.next(1)
      expect(callbacks).toMatchSnapshot()
      observable.next(2)
      expect(callbacks).toMatchSnapshot()
      observable.next(3)
      expect(callbacks).toMatchSnapshot()
      observable.complete()
      expect(callbacks).toMatchSnapshot()
    })
  })

  describe('errors', () => {
    it('should exponentially backoff and retry', () => {
      const mocks = {
        onNext: jest.fn(),
        onError: jest.fn(),
        onComplete: jest.fn(),
        subscribe: jest.fn(),
        optsOnError: jest.fn()
      }
      let failCount = 0
      const subscribe = (observer) => {
        mocks.subscribe(observer.constructor.name) // for snapshot
        observer.error(failCount)
        failCount++
        return new Subscription()
      }
      const observable = new Observable(subscribe)
      const opts = {
        // 5, 10, 20
        minTimeout: 5,
        maxTimeout: 10 * Math.pow(2, 2),
        factor: 2,
        retries: 2,
        onError: mocks.optsOnError
      }
      observable
        .backoff(opts)
        .subscribe(
          mocks.onNext,
          mocks.onError,
          mocks.onComplete
        )
      return timeout(0)
        .then(() => expect(mocks).toMatchSnapshot())
        .then(() => timeout(5 + 1))
        .then(() => expect(mocks).toMatchSnapshot())
        .then(() => timeout(10 + 1))
        .then(() => expect(mocks).toMatchSnapshot())
        .then(() => timeout(20 + 1)) // extra to ensure it doesn't retry too many times
        .then(() => expect(mocks).toMatchSnapshot())
    })
    it('should exponentially backoff and stop if onError returns true', () => {
      const mocks = {
        onNext: jest.fn(),
        onError: jest.fn(),
        onComplete: jest.fn(),
        subscribe: jest.fn(),
        optsOnError: jest.fn(function (err, final, count) {
          const dontRetry = count === 3
          return dontRetry
        })
      }
      let failCount = 0
      const subscribe = (observer) => {
        mocks.subscribe(observer.constructor.name) // for snapshot
        observer.error(failCount)
        failCount++
        return new Subscription()
      }
      const observable = new Observable(subscribe)
      const opts = {
        // 5, 10, 20
        minTimeout: 5,
        maxTimeout: 10 * Math.pow(2, 2),
        factor: 2,
        retries: 2,
        onError: mocks.optsOnError
      }
      observable
        .backoff(opts)
        .subscribe(
          mocks.onNext,
          mocks.onError,
          mocks.onComplete
        )
      return timeout(0)
        .then(() => expect(mocks).toMatchSnapshot())
        .then(() => timeout(5 + 1))
        .then(() => expect(mocks).toMatchSnapshot())
        .then(() => timeout(10 + 1)) // stopped retrying..
        .then(() => expect(mocks).toMatchSnapshot())
        .then(() => timeout(20 + 1))
        .then(() => expect(mocks).toMatchSnapshot())
    })
  })
})


index.js
'use strict'

var defaults = require('101/defaults')
var noop = require('101/noop')
var Observable = require('rxjs/Observable').Observable
var StaticObservable = require('static-observable')
// observable factories
require('rxjs/add/observable/range')
require('rxjs/add/observable/zip')
// operators
require('rxjs/add/operator/delay')
require('rxjs/add/operator/mergeMap')
require('rxjs/add/operator/retryWhen')

/**
 * exponential backoff operator
 * @return {[type]} [description]
 */
Observable.prototype.backoff = function (opts) {
  opts = opts || {}
  defaults(opts, {
    minTimeout: 1000,
    maxTimeout: 3000,
    factor: 3,
    retries: 2,
    onError: noop
  })
  return this.retryWhen(function (errorObs) {
    return Observable.zip(
      Observable.range(0, opts.retries + 1),
      errorObs
    ).mergeMap(function (zip) {
      var i = zip[0]
      var err = zip[1]
      var dontRetry = opts.onError(err, i === opts.retries, i + 1)
      if (dontRetry || i === opts.retries) {
        return StaticObservable.error(err)
      }
      var timeout = Math.min(
        opts.minTimeout * Math.pow(opts.factor, i),
        opts.maxTimeout
      )
      return StaticObservable.next(i).delay(timeout)
    })
  })
}
